File: clients/fsharp/src/Commands.fs
// SPDX-License-Identifier: GPL-2.0-only
//
// HighBarV3 — F# ergonomic wrapper over AICommand + SubmitCommands
// client-streaming helper (T062).
//
// The generated Highbar.V1.AICommand is a oneof over 97 command types;
// building the oneof by hand is verbose. This module exposes a small
// F# DU over the commonly-used unit-order commands and a batch helper
// that takes care of target_unit_id assignment, CommandOptions, and
// the client-stream ACK round trip (FR-010, FR-012a).
//
// Coverage mirrors CommandDispatch on the C++ side: move, patrol,
// fight, attack-area, stop, wait, build, repair, reclaim-unit,
// reclaim-area, resurrect-area, self-destruct, wanted-speed,
// fire-state, move-state. The generated AICommand.Builder is still
// accessible for clients that need the long tail.

namespace HighBar.Client

open System.Collections.Generic
open System.Threading
open System.Threading.Tasks
open Grpc.Core
open Grpc.Net.Client
open Highbar.V1

module Commands =

    /// Convenience type alias — the proto Vector3 is exposed to clients
    /// under Highbar.V1 but lifted here so call sites don't have to
    /// open the whole namespace.
    type Vec3 = Vector3

    let vec3 (x: float32) (y: float32) (z: float32) : Vec3 =
        Vector3(X = x, Y = y, Z = z)

    /// F# DU over the unit-order command set covered by the C++
    /// dispatcher. Add arms as the C++ side grows coverage — out-of-DU
    /// arms can still be issued by building AICommand directly.
    type Order =
        | MoveTo      of Vec3
        | PatrolTo    of Vec3
        | FightTo     of Vec3
        | AttackArea  of pos: Vec3 * radius: float32
        | Stop
        | Wait
        | Build       of defId: int32 * pos: Vec3 * facing: int32
        | Repair      of repairUnitId: int32
        | ReclaimUnit of reclaimUnitId: int32
        | ReclaimArea of pos: Vec3 * radius: float32
        | ResurrectArea of pos: Vec3 * radius: float32
        | SelfDestruct
        | WantedSpeed of float32
        | FireState   of int32
        | MoveState   of int32

    /// CommandOptions bitfield per common.proto. Use when the batch
    /// should queue (SHIFT) instead of interrupt-current.
    [<Struct>]
    type Opts =
        { Shift: bool
          Ctrl: bool
          Alt: bool }
        static member None = { Shift = false; Ctrl = false; Alt = false }
        static member Queued = { Shift = true; Ctrl = false; Alt = false }

    let private optsToBitfield (o: Opts) : uint32 =
        let mutable b = 0u
        if o.Shift then b <- b ||| 1u
        if o.Ctrl then b <- b ||| 2u
        if o.Alt then b <- b ||| 4u
        b

    /// Build an AICommand from an `Order` + the target unit id + options.
    let toProto (unitId: int32) (opts: Opts) (order: Order) : AICommand =
        let cmd = AICommand()
        let bits = optsToBitfield opts
        match order with
        | MoveTo pos ->
            cmd.MoveUnit <- MoveUnitCommand(
                UnitId = unitId, Options = bits, ToPosition = pos)
        | PatrolTo pos ->
            cmd.Patrol <- PatrolCommand(
                UnitId = unitId, Options = bits, ToPosition = pos)
        | FightTo pos ->
            cmd.Fight <- FightCommand(
                UnitId = unitId, Options = bits, ToPosition = pos)
        | AttackArea (pos, radius) ->
            cmd.AttackArea <- AttackAreaCommand(
                UnitId = unitId, Options = bits,
                AttackPosition = pos, Radius = radius)
        | Stop ->
            cmd.Stop <- StopCommand(UnitId = unitId, Options = bits)
        | Wait ->
            cmd.Wait <- WaitCommand(UnitId = unitId, Options = bits)
        | Build (defId, pos, facing) ->
            cmd.BuildUnit <- BuildUnitCommand(
                UnitId = unitId, Options = bits,
                ToBuildUnitDefId = defId,
                BuildPosition = pos, Facing = facing)
        | Repair repairId ->
            cmd.Repair <- RepairCommand(
                UnitId = unitId, Options = bits, RepairUnitId = repairId)
        | ReclaimUnit reclaimId ->
            cmd.ReclaimUnit <- ReclaimUnitCommand(
                UnitId = unitId, Options = bits, ReclaimUnitId = reclaimId)
        | ReclaimArea (pos, radius) ->
            cmd.ReclaimArea <- ReclaimAreaCommand(
                UnitId = unitId, Options = bits,
                Position = pos, Radius = radius)
        | ResurrectArea (pos, radius) ->
            cmd.ResurrectInArea <- ResurrectInAreaCommand(
                UnitId = unitId, Options = bits,
                Position = pos, Radius = radius)
        | SelfDestruct ->
            cmd.SelfDestruct <- SelfDestructCommand(
                UnitId = unitId, Options = bits)
        | WantedSpeed v ->
            cmd.SetWantedMaxSpeed <- SetWantedMaxSpeedCommand(
                UnitId = unitId, Options = bits, WantedMaxSpeed = v)
        | FireState s ->
            cmd.SetFireState <- SetFireStateCommand(
                UnitId = unitId, Options = bits, FireState = s)
        | MoveState s ->
            cmd.SetMoveState <- SetMoveStateCommand(
                UnitId = unitId, Options = bits, MoveState = s)
        cmd

    /// Build a CommandBatch carrying orders for a single target.
    let batch (targetUnitId: uint32)
              (batchSeq: uint64)
              (opts: Opts)
              (orders: Order list)
              : CommandBatch =
        let b = CommandBatch(
                    BatchSeq = batchSeq,
                    TargetUnitId = targetUnitId)
        for ord in orders do
            b.Commands.Add(toProto (int32 targetUnitId) opts ord)
        b

    /// SubmitCommands session wrapper. Opens the client-streaming RPC,
    /// writes each batch, and surfaces the CommandAck on completion.
    /// The AI token MUST have been added to `metadata` by the caller —
    /// Session.readTokenWithBackoff + manual Metadata.Add is the
    /// idiomatic path (see samples/AiClient).
    type SubmitSession(channel: GrpcChannel, metadata: Metadata, ct: CancellationToken) =
        let client = HighBarProxy.HighBarProxyClient(channel)
        let call = client.SubmitCommands(metadata, cancellationToken = ct)

        /// Send a CommandBatch. Throws RpcException on queue overflow
        /// (RESOURCE_EXHAUSTED) or validation failure (INVALID_ARGUMENT)
        /// — the server closes the stream on either.
        member _.SendAsync(batch: CommandBatch) : Task =
            call.RequestStream.WriteAsync(batch)

        /// Close the client-side of the stream and await the final
        /// CommandAck.
        member _.CompleteAsync() : Task<CommandAck> =
            task {
                do! call.RequestStream.CompleteAsync()
                return! call.ResponseAsync
            }

        interface System.IDisposable with
            member _.Dispose() = call.Dispose()

    /// Convenience: open a submit session, write a single batch, close,
    /// and return the ACK. Appropriate for fire-and-forget patterns
    /// (samples, tests). For long-lived streaming use SubmitSession.
    let submitOne (channel: GrpcChannel)
                  (metadata: Metadata)
                  (b: CommandBatch)
                  (ct: CancellationToken)
                  : Async<CommandAck> =
        async {
            use session = new SubmitSession(channel, metadata, ct)
            do! session.SendAsync(b) |> Async.AwaitTask
            let! ack = session.CompleteAsync() |> Async.AwaitTask
            return ack
        }

File: clients/python/highbar_client/commands.py
# SPDX-License-Identifier: GPL-2.0-only
"""Command helpers + SubmitCommands client stream (T089).

Mirrors the coverage in clients/fsharp/src/Commands.fs — 15 unit-order
arms. The raw AICommand proto is still accessible for the long tail.
"""

from __future__ import annotations

import enum
from collections.abc import Iterable, Iterator
from dataclasses import dataclass
from typing import Optional

from .highbar import (  # type: ignore
    commands_pb2,
    common_pb2,
    service_pb2,
)
TOKEN_HEADER = "x-highbar-ai-token"


def vec3(x: float, y: float, z: float) -> common_pb2.Vector3:
    return common_pb2.Vector3(x=x, y=y, z=z)


class OptionBits(enum.IntFlag):
    NONE = 0
    SHIFT = 1  # queued
    CTRL = 2
    ALT = 4


@dataclass(frozen=True, slots=True)
class Order:
    """Internal sum type — use the builder functions below."""

    kind: str
    payload: dict


def move_to(x: float, y: float, z: float) -> Order:
    return Order("move_to", {"pos": vec3(x, y, z)})


def patrol_to(x: float, y: float, z: float) -> Order:
    return Order("patrol_to", {"pos": vec3(x, y, z)})


def fight_to(x: float, y: float, z: float) -> Order:
    return Order("fight_to", {"pos": vec3(x, y, z)})


def attack_area(x: float, y: float, z: float, radius: float) -> Order:
    return Order("attack_area", {"pos": vec3(x, y, z), "radius": radius})


def stop() -> Order:
    return Order("stop", {})


def wait() -> Order:
    return Order("wait", {})


def build(def_id: int, x: float, y: float, z: float, facing: int = 0) -> Order:
    return Order(
        "build",
        {"def_id": def_id, "pos": vec3(x, y, z), "facing": facing},
    )


def repair(repair_unit_id: int) -> Order:
    return Order("repair", {"repair_unit_id": repair_unit_id})


def reclaim_unit(reclaim_unit_id: int) -> Order:
    return Order("reclaim_unit", {"reclaim_unit_id": reclaim_unit_id})


def reclaim_area(x: float, y: float, z: float, radius: float) -> Order:
    return Order("reclaim_area", {"pos": vec3(x, y, z), "radius": radius})


def resurrect_area(x: float, y: float, z: float, radius: float) -> Order:
    return Order("resurrect_area", {"pos": vec3(x, y, z), "radius": radius})


def self_destruct() -> Order:
    return Order("self_destruct", {})


def wanted_speed(speed: float) -> Order:
    return Order("wanted_speed", {"speed": speed})


def fire_state(state: int) -> Order:
    return Order("fire_state", {"state": state})


def move_state(state: int) -> Order:
    return Order("move_state", {"state": state})


def _to_proto(
    order: Order, unit_id: int, opts: OptionBits
) -> commands_pb2.AICommand:
    cmd = commands_pb2.AICommand()
    bits = int(opts)
    k = order.kind
    p = order.payload
    if k == "move_to":
        cmd.move_unit.unit_id = unit_id
        cmd.move_unit.options = bits
        cmd.move_unit.to_position.CopyFrom(p["pos"])
    elif k == "patrol_to":
        cmd.patrol.unit_id = unit_id
        cmd.patrol.options = bits
        cmd.patrol.to_position.CopyFrom(p["pos"])
    elif k == "fight_to":
        cmd.fight.unit_id = unit_id
        cmd.fight.options = bits
        cmd.fight.to_position.CopyFrom(p["pos"])
    elif k == "attack_area":
        cmd.attack_area.unit_id = unit_id
        cmd.attack_area.options = bits
        cmd.attack_area.attack_position.CopyFrom(p["pos"])
        cmd.attack_area.radius = p["radius"]
    elif k == "stop":
        cmd.stop.unit_id = unit_id
        cmd.stop.options = bits
    elif k == "wait":
        cmd.wait.unit_id = unit_id
        cmd.wait.options = bits
    elif k == "build":
        cmd.build_unit.unit_id = unit_id
        cmd.build_unit.options = bits
        cmd.build_unit.to_build_unit_def_id = p["def_id"]
        cmd.build_unit.build_position.CopyFrom(p["pos"])
        cmd.build_unit.facing = p["facing"]
    elif k == "repair":
        cmd.repair.unit_id = unit_id
        cmd.repair.options = bits
        cmd.repair.repair_unit_id = p["repair_unit_id"]
    elif k == "reclaim_unit":
        cmd.reclaim_unit.unit_id = unit_id
        cmd.reclaim_unit.options = bits
        cmd.reclaim_unit.reclaim_unit_id = p["reclaim_unit_id"]
    elif k == "reclaim_area":
        cmd.reclaim_area.unit_id = unit_id
        cmd.reclaim_area.options = bits
        cmd.reclaim_area.position.CopyFrom(p["pos"])
        cmd.reclaim_area.radius = p["radius"]
    elif k == "resurrect_area":
        cmd.resurrect_in_area.unit_id = unit_id
        cmd.resurrect_in_area.options = bits
        cmd.resurrect_in_area.position.CopyFrom(p["pos"])
        cmd.resurrect_in_area.radius = p["radius"]
    elif k == "self_destruct":
        cmd.self_destruct.unit_id = unit_id
        cmd.self_destruct.options = bits
    elif k == "wanted_speed":
        cmd.set_wanted_max_speed.unit_id = unit_id
        cmd.set_wanted_max_speed.options = bits
        cmd.set_wanted_max_speed.wanted_max_speed = p["speed"]
    elif k == "fire_state":
        cmd.set_fire_state.unit_id = unit_id
        cmd.set_fire_state.options = bits
        cmd.set_fire_state.fire_state = p["state"]
    elif k == "move_state":
        cmd.set_move_state.unit_id = unit_id
        cmd.set_move_state.options = bits
        cmd.set_move_state.move_state = p["state"]
    else:
        raise ValueError(f"unknown order kind: {k}")
    return cmd


def batch(
    target_unit: int,
    batch_seq: int,
    orders: Iterable[Order],
    opts: OptionBits = OptionBits.NONE,
    client_command_id: int | None = None,
    based_on_frame: int | None = None,
    based_on_state_seq: int | None = None,
    conflict_policy: int | None = None,
) -> commands_pb2.CommandBatch:
    b = commands_pb2.CommandBatch(batch_seq=batch_seq, target_unit_id=target_unit)
    if client_command_id is not None:
        b.client_command_id = client_command_id
    if based_on_frame is not None:
        b.based_on_frame = based_on_frame
    if based_on_state_seq is not None:
        b.based_on_state_seq = based_on_state_seq
    if conflict_policy is not None:
        b.conflict_policy = conflict_policy
    for ord_ in orders:
        b.commands.append(_to_proto(ord_, target_unit, opts))
    return b


def issue_summary(
    result: commands_pb2.CommandBatchResult,
) -> list[tuple[str, int, str, str]]:
    return [
        (
            commands_pb2.CommandIssueCode.Name(issue.code),
            issue.command_index,
            issue.field_path,
            commands_pb2.RetryHint.Name(issue.retry_hint),
        )
        for issue in result.issues
    ]


def submit_one(
    channel_,
    token: str,
    batch_: commands_pb2.CommandBatch,
    timeout: Optional[float] = None,
) -> service_pb2.CommandAck:
    """One-shot SubmitCommands: send a single batch and await the ack."""

    def _batches() -> Iterator[commands_pb2.CommandBatch]:
        yield batch_

    from .highbar import service_pb2_grpc  # type: ignore

    stub = service_pb2_grpc.HighBarProxyStub(channel_)
    metadata = [(TOKEN_HEADER, token)]
    return stub.SubmitCommands(_batches(), metadata=metadata, timeout=timeout)


def validate_batch(
    channel_,
    token: str,
    batch_: commands_pb2.CommandBatch,
    timeout: Optional[float] = None,
) -> commands_pb2.CommandBatchResult:
    from .highbar import service_pb2_grpc  # type: ignore

    stub = service_pb2_grpc.HighBarProxyStub(channel_)
    metadata = [(TOKEN_HEADER, token)]
    return stub.ValidateCommandBatch(batch_, metadata=metadata, timeout=timeout)
